Refactor Maven publication: maven -> maven-publish

vlsi opened the pull request:
See https://github.com/GoogleCloudPlatform/appengine-plugins/issues/1004

This uses Gradle's default means for generating plugin markers.

You can publish the artifacts to MavenLocal repository with `./gradlew publishToMavenLocal`

As an alternative option, you can run `./gradlew publishToTmpMaven` to publish the files to `build/repo` directory.

```
$ rm -rf build/repo
$ ./gradlew -Pversion=2.4.6 publishToTmpMaven
```

=>

```
$ tree -I '*.md5|*.sha*|maven-metadata.xml' build/repo

build/repo
└── com
    └── google
        └── cloud
            └── tools
                ├── appengine
                │   └── com.google.cloud.tools.appengine.gradle.plugin
                │       └── 2.4.6
                │           └── com.google.cloud.tools.appengine.gradle.plugin-2.4.6.pom
                ├── appengine-appenginewebxml
                │   └── com.google.cloud.tools.appengine-appenginewebxml.gradle.plugin
                │       └── 2.4.6
                │           └── com.google.cloud.tools.appengine-appenginewebxml.gradle.plugin-2.4.6.pom
                ├── appengine-appyaml
                │   └── com.google.cloud.tools.appengine-appyaml.gradle.plugin
                │       └── 2.4.6
                │           └── com.google.cloud.tools.appengine-appyaml.gradle.plugin-2.4.6.pom
                ├── appengine-flexible
                │   └── com.google.cloud.tools.appengine-flexible.gradle.plugin
                │       └── 2.4.6
                │           └── com.google.cloud.tools.appengine-flexible.gradle.plugin-2.4.6.pom
                ├── appengine-gradle-plugin
                │   └── 2.4.6
                │       ├── appengine-gradle-plugin-2.4.6-javadoc.jar
                │       ├── appengine-gradle-plugin-2.4.6-sources.jar
                │       ├── appengine-gradle-plugin-2.4.6.jar
                │       ├── appengine-gradle-plugin-2.4.6.module
                │       └── appengine-gradle-plugin-2.4.6.pom
                ├── appengine-standard
                │   └── com.google.cloud.tools.appengine-standard.gradle.plugin
                │       └── 2.4.6
                │           └── com.google.cloud.tools.appengine-standard.gradle.plugin-2.4.6.pom
                └── source-context
                    └── com.google.cloud.tools.source-context.gradle.plugin
                        └── 2.4.6
                            └── com.google.cloud.tools.source-context.gradle.plugin-2.4.6.pom
```

----

There's `./gradlew prepareRelease` task to mimic the previous behaviour:

```
$ rm -rf build/release-artifacts
$ ./gradlew -Pversion=2.4.6 prepareRelease
```

=>

```
$ tree -P '*' build/release-artifacts
build/release-artifacts
├── plugin-artifacts
│   ├── appengine-gradle-plugin-2.4.6-javadoc.jar
│   ├── appengine-gradle-plugin-2.4.6-sources.jar
│   ├── appengine-gradle-plugin-2.4.6.jar
│   ├── appengine-gradle-plugin-2.4.6.module
│   └── appengine-gradle-plugin-2.4.6.pom
└── plugin-markers
    ├── com.google.cloud.tools.appengine-appenginewebxml.gradle.plugin-2.4.6.pom
    ├── com.google.cloud.tools.appengine-appyaml.gradle.plugin-2.4.6.pom
    ├── com.google.cloud.tools.appengine-flexible.gradle.plugin-2.4.6.pom
    ├── com.google.cloud.tools.appengine-standard.gradle.plugin-2.4.6.pom
    ├── com.google.cloud.tools.appengine.gradle.plugin-2.4.6.pom
    └── com.google.cloud.tools.source-context.gradle.plugin-2.4.6.pom
```

vlsi commented:
@loosebazooka , I believe the PR is ready. It generates the plugin markers and publishes them in one go (see sample  https://oss.sonatype.org/content/repositories/comgithubvlsi-1131/ )

loosebazooka commented:
unfortunately, the workflow to do
```
signing.gnupg.keyName=AC2AEE0F
signing.keyId=AC2AEE0E
signing.password=...
signing.secretKeyRingFile=/path/to/secring.gpg
```
is not viable given our signing process.

vlsi commented:
What do you do for signing?
Do you want to sign and upload all the artifacts manually?

vlsi commented:
Just in case, all the above options can be passed with environment variables and/or with gpg-agent.

loosebazooka commented:
So our signing machine doesn't have internet access.

The artifact is created -> copied to signing vm -> signed -> copied to another vm -> uploaded to sonatype.

There's not much more I can say about the signing process. However, I'll play around with this and see if I can make it work. This is definitely an improvement that is useful to users.

loosebazooka commented:
so yeah as long as you can generate all the necessary files into a known target directory, we can handle the followup work to publish it.

vlsi commented:
What do you think of the following?

```
$ rm -rf build/repo
$ ./gradlew -Pversion=2.4.3 publish
```

=>

```
tree -P '*.pom|*.jar' build/repo
build/repo
└── com
    └── google
        └── cloud
            └── tools
                ├── appengine
                │   └── com.google.cloud.tools.appengine.gradle.plugin
                │       └── 2.4.3
                │           └── com.google.cloud.tools.appengine.gradle.plugin-2.4.3.pom
                ├── appengine-appenginewebxml
                │   └── com.google.cloud.tools.appengine-appenginewebxml.gradle.plugin
                │       └── 2.4.3
                │           └── com.google.cloud.tools.appengine-appenginewebxml.gradle.plugin-2.4.3.pom
                ├── appengine-appyaml
                │   └── com.google.cloud.tools.appengine-appyaml.gradle.plugin
                │       └── 2.4.3
                │           └── com.google.cloud.tools.appengine-appyaml.gradle.plugin-2.4.3.pom
                ├── appengine-flexible
                │   └── com.google.cloud.tools.appengine-flexible.gradle.plugin
                │       └── 2.4.3
                │           └── com.google.cloud.tools.appengine-flexible.gradle.plugin-2.4.3.pom
                ├── appengine-gradle-plugin
                │   └── 2.4.3
                │       ├── appengine-gradle-plugin-2.4.3-javadoc.jar
                │       ├── appengine-gradle-plugin-2.4.3-sources.jar
                │       ├── appengine-gradle-plugin-2.4.3.jar
                │       └── appengine-gradle-plugin-2.4.3.pom
                ├── appengine-standard
                │   └── com.google.cloud.tools.appengine-standard.gradle.plugin
                │       └── 2.4.3
                │           └── com.google.cloud.tools.appengine-standard.gradle.plugin-2.4.3.pom
                └── source-context
                    └── com.google.cloud.tools.source-context.gradle.plugin
                        └── 2.4.3
                            └── com.google.cloud.tools.source-context.gradle.plugin-2.4.3.pom
```
```

loosebazooka commented:
So again, thnks for your interest in doing this, and sorry that our release is a little bit of a blackbox.

Right now what we do is generate a special folder "release-artifacts" that contains everything we need. This helps remove the logic of parsing the maven output from the release process.

Could we do something similar to the current `prepareRelease` task? and potentially create the following output dirs?

```
- release-artifacts
 - plugin-artifacts
   - appengine-gradle-plugin-2.4.3-javadoc.jar
   - appengine-gradle-plugin-2.4.3-sources.jar
   - appengine-gradle-plugin-2.4.3.jar
   - appengine-gradle-plugin-2.4.3.pom
 - plugin-markers 
   - com.google.cloud.tools.appengine-standard.gradle.plugin-2.4.3.pom
   - com.google.cloud.tools.source-context.gradle.plugin-2.4.3.pom
   - etc.
```

chanseokoh commented:
Thanks for the interest! Can we retain the Groovy DSL syntax? Not that I don't like Kotlin, but I have zero knowledge with Kotlin and previously I had a lot of trouble fixing Gradle issues when dealing with Kotlin DSL (yeah, my problem). Even most of the team members don't have great expertise with Gradle, so I am afraid this may have adverse impact on us.

vlsi commented:
> Could we do something similar to the current prepareRelease task? and potentially create the following output dirs?

I guess it can be done, however, it would require non-standard trickery. I am afraid it might be hard to maintain in the future.

Are you sure the standard Maven layout does not work for you?
The above `build/repo` layout is standard, it is predictable, and it would automatically handle new artifacts in case you add them in the future. It is documented in Gradle documentation.

----

> Can we retain the Groovy DSL syntax? 

Ah. Let me see. I tried Groovy DSL, and it is hard to write for me as autocomplete does not really work there.
Of course, I could convert this script to Groovy, however, that would be really annoying for me.

> Even most of the team members don't have great expertise with Gradle,

Well, that might depend on the IDE, however, I really got comfortable with Gradle only when I tried Kotlin DSL.
Then it does help if the plugin code and the build scripts are written in Kotlin: the same language makes it easier to reason about and copy&paste code )

vlsi commented:
@loosebazooka , I've added `./gradlew prepareRelease` that shuflfes output files to `build/release-artifacts/plugin-artifacts` and `build/release-artifacts/plugin-markers `

martinbonnin commented:
Any blockers merging this? Having plugin markers available would make it easier to use the plugin.

meltsufin commented:
@martinbonnin This PR is blocked on reverting to Groovy DSL.

martinbonnin commented:
I made a very crude first attempt at reverting to Groovy there: https://github.com/vlsi/app-gradle-plugin/pull/1

@meltsufin any chance we can convince you to stick with Kotlin build scripts before reverting everything? Having autocomplete in the IDE is really nice. There are definitely downsides such as bigger compilation times but I find the developer experience much better.

vlsi commented:
@jamesward , please, https://twitter.com/_JamesWard/status/1472961164724289537 🙃 

WDYT on https://github.com/GoogleCloudPlatform/app-gradle-plugin/pull/398#issuecomment-1005158529 ? 😉

chanseokoh commented:
@vlsi the team has gained sufficient expertise with Kotlin, so we can actually accept the Kotlin migration. 

However, please keep the Groovy DSL in this PR. Fixing #397 and migrating to Kotlin at the same time gives hard time to reviewers, and if we ever have to revert either change, it is going to be a lot more work compared to filing two separate PRs. Big migration like this one has to be done in its own PR.

vlsi commented:
@meltsufin, I see, thank you. I would split the PR unless Martin makes it faster.

martinbonnin commented:
Go for it, I have a plentifull beginning of 2022!

chanseokoh commented:
FTR, status: draft

martinbonnin commented:
I gave this another shot with https://github.com/GoogleCloudPlatform/app-gradle-plugin/pull/423. #423 focuses at moving the build scripts to Kotlin. Let me know how that sounds.

elefeint commented:
Thanks! Smaller PR is definitely easier.

vlsi commented:
@chanseokoh, this is not draft, and I believe it is ready to review and merge.

martinbonnin commented:
@vlsi want to rebase now that scripts are in Kotlin (see #423)?

loosebazooka commented on build.gradle.kts:
```diff
@@ -0,0 +1,329 @@
+            description =
+                "This Gradle plugin provides tasks to build and deploy Google App Engine applications."
+        }
+        create("appengine-appenginewebxml") {
+            id = "$group.appengine-appenginewebxml"
+            implementationClass =
+                "com.google.cloud.tools.gradle.appengine.standard.AppEngineStandardPlugin"
+            displayName = "App Engine Gradle Plugin"
+            description =
+                "This Gradle plugin provides tasks to build and deploy Google App Engine applications."
+        }
+        create("appengine-appyaml") {
+            id = "$group.appengine-appyaml"
+            implementationClass =
+                "com.google.cloud.tools.gradle.appengine.appyaml.AppEngineAppYamlPlugin"
+            displayName = "App Engine Gradle Plugin"
+            description =
+                "This Gradle plugin provides tasks to build and deploy Google App Engine applications."
```
I think these would be the only two we actually want to publish to maven central. All other naming is deprecated?

vlsi replied:
I kept all the plugins, just like they were declared in properties files.
If you want to drop some, that is doable, however, I've no idea which ones are used in the wild.

loosebazooka replied:
Hah yeah they're all in the wild. Many years ago, it was proposed to publish a subset of these to the gradle-plugin portal. **Maybe it's not worth dealing with here**. But the last major user guide update has some info here: https://github.com/GoogleCloudPlatform/app-gradle-plugin/blob/master/USER_GUIDE.md#user-guide-2

Perhaps this can be dealt with after the refactor.

vlsi replied:
Just in case, after this refactor, the plugin will work via `plugins { id(...) }` even if you publish to Central only (I know Google did not like to publish to Portal for some reason).

loosebazooka replied:
@elfel for future reference, a lot of docs probably still point to "appengine" so maybe we keep 3... appengine, appengine-appyaml and appengine-appenginewebxml.

vlsi commented:
The PR is ready for review.

Note: the PR includes 8 commits, and they can be reviewed and merged individually.

loosebazooka commented:
> Just in case, after this refactor, the plugin will work via plugins { id(...) } even if you publish to Central only (I know Google did not like to publish to Portal for some reason).

Yeah I tried it here, but it had some issues (which I can't remember anymore)

vlsi commented:
Anyway, I'm inclined that adding "publishing to Gradle Plugin Portal" is out of the scope of this PR

elefeint commented:
I'll try to review Friday.

elefeint reviewed (changes requested):
The individual commits were very helpful, thank you.

I am going to have to update the internal release configuration to be able to test staging, but looking at what gets generated by the current script, I can work with it.

vlsi commented on .editorconfig:
```diff
@@ -0,0 +1,16 @@
+root = true
```
Why's that?
editorconfig is helpful (e.g. "trim whitespace", "configure indentation", and even "import order for java"), so why do you suggest removing it?

elefeint replied:
This isn't the hill I am planning to die on, but the answer is because nobody on the team that maintains this project uses editorconfig plugins.

TWiStErRob replied:
Do you use IntelliJ IDEA to edit the code, @elefeint? If so, you are using EditorConfig plugins: https://www.jetbrains.com/help/idea/editorconfig.html

Note that checking in this file would help **external contributors** (e.g. I edited the code, pressed autoformat, and EVERYTHING changed). We could have 0-configuration experience (import and it "just works"). Discovering and running `gradlew googleJavaFormat` is just not the norm in the world of IDEs outside of Google.

vlsi commented on build.gradle.kts:
```diff
@@ -18,197 +14,296 @@ import org.gradle.api.tasks.testing.logging.TestExceptionFormat.FULL
-  id("com.github.sherter.google-java-format") version "0.8"
-  id("checkstyle")
-  id("jacoco")
+    id("java")
```
Can you please point me to Google style guide for Kotlin that says "2 spaces"?

https://github.com/google/styleguide/issues/314 says Google style is (see https://github.com/google/styleguide/pull/365/files) https://developer.android.com/kotlin/style-guide, which says "4 spaces" https://developer.android.com/kotlin/style-guide#indentation just like the Kotlin official coding conventions: https://kotlinlang.org/docs/coding-conventions.html#indentation

vlsi replied:
@elefeint , would you please clarify if "squashing on merge" is an absolute requirement?
Squashing commits breaks git history analysis, and committing individual changes as separate commits helps that.

Do you suggest I should file 8 individual PRs?

The squash in https://github.com/GoogleCloudPlatform/app-gradle-plugin/pull/423#issuecomment-1242909704 caused `build.gradle.kts` to look like "a completely fresh file written by @martinbonnin" even though Martin tried their best to translate to Kotlin one-to-one.

----

> More importantly, keeping unrelated style changes out of PRs makes the inevitable code archaeology so much easier

In fact, IntelliJ IDEA shows annotate for the "re-indented" lines just fine:

<img width="798" alt="git annotate in IDEA for build.gradle.kts" src="https://user-images.githubusercontent.com/213894/189517629-e2c4287f-86d0-40e3-b499-3c35ca9d06d0.png">

As you see, the indentation is new, however, "commit line" attribution is proper (it does not show all the lines as if they were modified by me.

GitHub UI is slightly different, and it does treat re-indent as commit that changes the line: https://github.com/vlsi/app-gradle-plugin/blame/refactor_publication/build.gradle.kts#L61, however, if the changes are NOT squashed, then there's a convenient "show changes prior the commit" button that shows the annotations prior to re-indent: https://github.com/vlsi/app-gradle-plugin/blame/2a49f56b03fa5c71343f829efd189debb3b9bcc3/build.gradle.kts

I incline that committing individual changes as separate commits helps for `code archaeology`, and re-indenting as a separate commit does not make it much harder.

elefeint replied:
Squashing is not an absolute requirement; merely a common org/project practice. Since your individual commits are well structured, I can rebase-merge this pull request specifically.
For the indentation, though, keeping 2 space indents would be ideal. We will work on externalizing Google's Kotlin style guide with @jamesward; in the meantime you'll have to take my word for it, I'm afraid.

vlsi commented on build.gradle.kts:
```diff
@@ -18,197 +14,296 @@ import org.gradle.api.tasks.testing.logging.TestExceptionFormat.FULL
+                inceptionYear.set("2016")
+                licenses {
+                    license {
+                        name.set("The Apache License, Version 2.0")
```
Are you sure you want `The Apache Software License, Version 2.0` here?

Note that the official license text at https://www.apache.org/licenses/LICENSE-2.0.txt does not mention "software license", it has title of "Apache License Version 2.0" (without software).
The same goes for the license headers in the project: https://github.com/GoogleCloudPlatform/app-gradle-plugin/blob/f50be8dcdde14ec6ef94947f1e21f5ea855be792/src/main/java/com/google/cloud/tools/gradle/appengine/AppEnginePlugin.java#L4
They mention `Licensed under the Apache License, Version 2.0` rather than `Apache Software License`.

SPDX references the license as `Apache License 2.0` rather than `Apache Software License 2.0`: https://spdx.org/licenses/Apache-2.0.html

elefeint replied:
I was trying to avoid changing the existing state as much as possible, but you are right.

vlsi commented on build.gradle.kts:
```diff
@@ -18,197 +14,296 @@ import org.gradle.api.tasks.testing.logging.TestExceptionFormat.FULL
+                "com.google.cloud.tools.gradle.appengine.sourcecontext.SourceContextPlugin"
+            displayName = "App Engine Gradle Plugin"
+            description =
+                "This Gradle plugin provides tasks to build and deploy Google App Engine applications."
```
Sure. If you could suggest the descriptions for other plugins, I could integrate them. However, previous code used the same description for all the plugins, and I've just no clue what all the plugins do

vlsi commented:
@elefeint , could you please review the PR again? I've rebased the PR and I added a couple of changes:
* `chore: simplify options.compilerArgs configuration`
* `chore: use compileOnly for gradleApi` (see https://github.com/gradle/gradle/pull/24286/files#diff-c6c21698523d8bd69ea22657e010eaf36ac20c33c18e5a45b4514c8bb58c0461R29 )

`./gradlew test` passes
`./gradlew integTest` fails like `Execution failed for task ':appengineDeploy'. > Project was not found in gcloud config`, however, I assume Gradle configs work, and the test fails because I miss gcloud configuration.

TWiStErRob commented:
@vlsi can you please address https://github.com/GoogleCloudPlatform/app-gradle-plugin/pull/398#discussion_r984944216 GitHub PR changes look like a mess, but they don't have to be :) (I know there are individual commits, but the changes added together should also just address what it says on the tin (title), right?)

Also since this is a major publishing change, you should compare the output of the published artifacts (e.g. same POM, same jar contents), tests passing might not be enough for making sure users are not breaking, or losing non-functional features (like browsing source code or javadoc, or getting the right licensing info from the POM).

elefeint commented:
@vlsi I've abdicated my GCP review powers in favor of building a data analytics platform.
@meltsufin can route this to the right person.

martinbonnin commented:
@meltsufin any news here? This PR adds a lot of nice improvements and makes it easier for newcomers to consume the plugin, would be really nice to merge it.

vlsi commented:
I've realigned the indentation to 2 spaces, however, it would be nice to see the indentation size documented somewhere.

>I know there are individual commits, but the changes added together should also just address what it says on the tin (title)

@TWiStErRob , I have no interest in splitting this PR into individual PRs as there's been virtually no feedback on the PR during the past couple of years.

TWiStErRob commented on build.gradle.kts:
```diff
@@ -71,107 +73,208 @@ tasks.jar.configure {
+  from(tempRepoPath) {
+    include("**/*.gradle.plugin/**/*.pom")
+    // Flatten hierarchy
+    eachFile {
+      path = "plugin-markers/$name"
+    }
```
Are you sure this flattening is required to separate the artifacts? They all have unique names.

vlsi replied:
I’ve no idea on what are the requirements

TWiStErRob commented on build.gradle.kts:
```diff
@@ -71,107 +73,208 @@ tasks.jar.configure {
+// </editor-fold>
+
+// <editor-fold defaultstate="expanded" desc="Releasing">
+val generatePom by tasks.registering {
+  description =
+    "Generates all the pom files to ${buildDir.name}/publications/ for manual inspection"
+  group = "release"
+  dependsOn(tasks.withType<GenerateMavenPom>())
+}
```
I think this is not necessary as one can just run `gradlew prepareRelease` and get an even better view, or browse the raw the maven repo as a result of `gradlew publish`.

vlsi replied:
There was generatePom task befor this PR, so I keep it for backward compatibility

TWiStErRob replied:
No, there was a `writePom` task that is now replaced with `generatePomFileForPluginMavenPublication`.

TWiStErRob commented on build.gradle.kts:
```diff
@@ -71,107 +73,208 @@ tasks.jar.configure {
-  into("${buildDir}/release-artifacts")
-
-  dependsOn(tasks.build)
+val prepareRelease by tasks.registering(Sync::class) {
```
I also noticed you removed the `build` task dependency, which I guess is a good move, unless there are manual release scripts which just call `prepareRelease`. Those might be less protected now if they assume that dependency.

TWiStErRob commented on build.gradle.kts:
```diff
@@ -71,107 +73,208 @@ tasks.jar.configure {
+        }
+      }
+      pom {
+        // <editor-fold defaultstate="collapsed" desc="Default POM values">
```
These regions are awesome, they're much better than the raw comments, or even `region`/`endregion` pair for IDEA. Where does the support for this in IDEA come from?

TWiStErRob commented on build.gradle.kts:
```diff
@@ -71,107 +73,208 @@ tasks.jar.configure {
+      // that were used in Gradle build/test.
+      versionMapping {
+        usage(Usage.JAVA_RUNTIME) {
+          fromResolutionResult()
```
`<scope>test</scope>` dependencies are gone from POM, good riddance.

TWiStErRob reviewed (approved):
I (an external contributor like you) reviewed the Gradle changes, they look good, this is mergeable if Kokoro can handle the directory change. I'll leave the .editorconfig up to the owners.

I like the small (safe) modernizations you did for the Gradle script. My comments are mostly small improvements to DX and maintainability. @vlsi feel free to do as you wish with them. 

I tested the output artifacts, they look equivalent to current `master` except the snapshot names, see comment on relevant line in `prepareRelease`.

> virtually no feedback on the PR during the past couple of years.

I feel you, it's strange that there are contributors working on stuff, but there's no activity from the owners (not just in this org/repo). While running `prepareRelease` I noticed a lot of javadoc warnings, I was thinking I could fix those, and then I realized I already fixed them a few months ago in #452...

> I have no interest in splitting this PR into individual PRs

I see now that it's not that many extra things, I think it was more in the past. I generally try to split different aspects into different PRs, because there's a higher chance of review when
 * the PR is smaller (e.g. `tasks.jar.configure {` -> `tasks.jar {`-like changes in a separate PR is a trivial review and merge),
 * or less contentious (e.g. .editorconfig might block this PR, even though everything else could be merged).
 * and both the above make the main changes more understandable and the PR smaller as well.

vlsi commented on build.gradle.kts:
```diff
@@ -71,107 +73,208 @@ tasks.jar.configure {
+      id = "$group.appengine-standard"
+      implementationClass =
+        "com.google.cloud.tools.gradle.appengine.standard.AppEngineStandardPlugin"
+      displayName = "App Engine Gradle Plugin"
+      description =
+        "This Gradle plugin provides tasks to build and deploy Google App Engine applications."
```
I think I copied the values from the old properties. Of course, adding better descriptions would be nice, however, it would increase the PR surface for no reason

vlsi commented on build.gradle.kts:
```diff
@@ -71,107 +73,208 @@ tasks.jar.configure {
+val prepareRelease by tasks.registering(Sync::class) {
+  description = "Copy release artifacts to ${buildDir.name}/release-artifacts/"
+  group = "release"
+  dependsOn(tasks.publish)
```
I think when publishing a release, the users should override the version to non-snapshot

TWiStErRob commented on build.gradle.kts:
```diff
@@ -71,107 +73,208 @@ tasks.jar.configure {
+      description =
+        "This Gradle plugin provides tasks to build and deploy Google App Engine applications."
+    }
+    create("appengine-appenginewebxml") {
```
Don't work against Gradle, work with it 😅. They're listed and actually useful if you just want to run that one task, and also to see what publications are there without having to touch the build scripts. These tasks helped me a lot while I was working on publishing multiple times.

TWiStErRob replied:
Note: having nicer names won't increase the complexity of the PR, but provide a better experience for anyone looking at tasks / reading build scripts.

martinbonnin commented:
@JoeWang1127 can you share more context why this was closed?

JoeWang1127 commented:
Hi @martinbonnin, we plan to archive this repo and I see this PR hasn't been touched for a few months.

The content of this repo now lives in https://github.com/GoogleCloudPlatform/appengine-plugins, feel free to open another one there.
